lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -2135,6 +2189,264 @@ LogicalResult WhileOp::canonicalize(WhileOp whileOp,
+}
+
+ParseResult InvokeOp::parse(OpAsmParser &parser, OperationState &result) {
+  StringAttr componentName;
+  SmallVector<OpAsmParser::UnresolvedOperand, 4> ports;
+  SmallVector<OpAsmParser::UnresolvedOperand, 4> inputs;
+  SmallVector<Attribute> portNames;
+  SmallVector<Attribute> inputNames;
+  SmallVector<Type, 4> types;
+  if (parser.parseSymbolName(componentName))
+    return failure();
+  FlatSymbolRefAttr callee = FlatSymbolRefAttr::get(componentName);
+  SMLoc loc = parser.getCurrentLocation();
+  result.addAttribute("callee", callee);
+  if (parseParameterList(parser, result, ports, inputs, portNames, inputNames,
+                         types))
+    return failure();
+  if (parser.resolveOperands(ports, types, loc, result.operands))
+    return failure();
+  if (parser.resolveOperands(inputs, types, loc, result.operands))
+    return failure();
+  result.addAttribute("portNames",
+                      ArrayAttr::get(parser.getContext(), portNames));
+  result.addAttribute("inputNames",
+                      ArrayAttr::get(parser.getContext(), inputNames));
+  return success();
+}
+
+void InvokeOp::print(OpAsmPrinter &p) {
+  // print the parameter list.
+  p << " @" << getCallee() << "(";
+  auto ports = this->getPorts();
+  auto inputs = this->getInputs();
+  for (size_t i = 0; i != ports.size(); i++) {
+    p << ports[i] << " = " << inputs[i];
+    if (i + 1 != ports.size())
+      p << ", ";
+  }
+  p << ") -> (";
+  // Print argument type.
+  for (size_t i = 0; i != ports.size(); i++) {
+    p << ports[i].getType();
+    if (i + 1 != ports.size())
+      p << ", ";
+  }
+  p << ")";
+}
+
+static LogicalResult verifyInvokeOpValue(InvokeOp op, Value value,
+                                         bool isDestination) {
+  bool isSource = !isDestination;
+  if (isPort(value))
+    return verifyPortDirection(op, value, isDestination);
+
+  if (isSource) {
+    // Refer to the above function verifyNotComplexSource for its role.
+    Operation *operation = value.getDefiningOp();
+    if (operation == nullptr)
+      return success();
+    if (auto *dialect = operation->getDialect();
+        isa<comb::CombDialect>(dialect))
+      return op->emitOpError("has source that is not a port or constant. "
+                             "Complex logic should be conducted in the guard.");
+  }
+  return success();
+}
+
+// Get the value of the Go port within an instance.
+Value InvokeOp::getInstGoValue() {
+  ComponentOp componentOp = (*this)->getParentOfType<ComponentOp>();
+  Operation *operation = componentOp.lookupSymbol(getCallee());
+  Value ret = nullptr;
+  // Get the value of the writer_en port.
+  if (isa<RegisterOp>(operation))
+    ret = operation->getResult(1);
+  else if (isa<MemoryOp, DivSPipeLibOp, DivUPipeLibOp, MultPipeLibOp,
+               RemSPipeLibOp, RemUPipeLibOp>(operation)) {
+    // Get the value of the writer_en port or go port.
+    ret = operation->getResult(2);
```

Comment:
Instead of all these nested `if else if ... else if`, I'd suggest just returning early in each case, i.e., here `return operation->getResult(2);`